iroh-net/src/hp/portmapper.rs

```diff
@@ -12,43 +30,435 @@ pub struct ProbeResult {
     pub upnp: bool,
 }
 
-/// A port mapping client.
-#[derive(Default, Debug, Clone)]
-pub struct Client {}
+#[derive(Debug)]
+enum Message {
+    /// Request to update the local port.
+    ///
+    /// The resulting external address can be obtained subscribing using the [`Handle`].
+    /// A value of `None` will deactivate port mapping.
+    UpdateLocalPort { local_port: Option<NonZeroU16> },
+    /// Request to probe the port mapping protocols.
+    ///
+    /// The requester should wait for the result at the [`oneshot::Receiver`] counterpart of the
+    /// [`oneshot::Sender`].
+    Probe {
+        result_tx: oneshot::Sender<Result<ProbeResult>>,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub struct Client {
+    /// A watcher over the most recent external address obtained from port mapping.
+    port_mapping: watch::Receiver<Option<SocketAddrV4>>,
+    /// Channel used to communicate with the port mapping service.
+    service_tx: mpsc::Sender<Message>,
+    /// A handle to the service that will cancel the spawned task once the client is dropped.
+    service_handle: std::sync::Arc<util::CancelOnDrop>,
+}
 
 impl Client {
-    pub fn new() -> Self {
-        Self::default()
+    /// Create a new port mapping client.
+    pub async fn new(local_port: Option<NonZeroU16>) -> Self {
+        let (service_tx, service_rx) = mpsc::channel(4);
+        let (current_mapping_tx, current_mapping_rx) = watch::channel(None);
+
+        let service = Service {
+            local_port,
+            last_upnp_gateway_addr: None,
+            rx: service_rx,
+            current_external_address: current_mapping_tx,
+        };
+
+        let handle =
+            util::CancelOnDrop::new("portmap_service", tokio::spawn(async {}).abort_handle());
```

Comment:
Is this like for later, or what exactly is the point of this empty `async {}`block?